### ISSUE TITLE
Meson build system

### ISSUE THREAD

**Issue by guijan** [JLErvin/berry#144]
Hello, I think currently it can be said berry runs its own build system. It's an original mixture of make and shell scripting that imitates autotools.

Would the project be open to adding [Meson](https://mesonbuild.com/)?

Offhand, these are some of its immediately relevant features:
- Portable
The current build system is not very portable. It seems to depend on GNU extensions to utilities like sed. Portability is an inherent part of Meson, no careful programming would be required to make it build unmodified on all the free unices.
- Integrated tests
Currently, testing berry would mean writing an original test framework to go with the original build system. With Meson, it's trivial to write shell scripts and C programs to run as tests and to run them in parallel as well as measure how much of the code gets run during testing.
- Simple
I haven't written Meson build definitions for berry yet, but berry is a simple program with 3 .c files (Meson takes care of headers automatically) and dependencies found through pkg-config. I doubt the Meson build definitions will come anywhere near to the 317 lines that ```Makefile``` and ```configure``` add up to.
- Automatic compile_commands.json
I see a compile_commands.json in examples/. Meson generates one automatically when you run "meson setup build," its equivalent of ./configure. The current one seems to be specific to the lead developer's machine, too.
- Implicit behavior
With Meson, a "make dist" target is not needed. Meson build definitions include enough information that Meson knows what to include in a distfile.
It's also not necessary to create "release" and "debug" targets, those are automatically created by Meson.
Meson automatically makes use of ccache if it's installed.
Meson automatically figures out the directory to install to, that information is included within Meson in every platform it runs on.

If you think build systems are a pain, you'll find Meson is a joy to work with.

**Comment by JLErvin** [comment]
This is probably a good discussion to have with @msharov, who implemented `berry`'s current build system.

As the maintainer of the project, I will say that I'm a bit hesitant to switch build systems, for a couple of reasons. For our current needs, the build system seems to work fine and seldom requires any changes. Switching build systems is likely to add extra work for package maintainers and might break some package builds while we wait for them to update things on their end (for example, I maintain the AUR package and I would imagine this would require changes there. I'm quite busy and not certain that this is a problem I want to spend a ton of time on).

I'd be interested in seeing what a protoype of this would look like to compare the complexity and see if it's worth investing time into.

**Comment by msharov** [comment]
I would recommend against it. Meson, like many other fads, is merely a solution looking for a problem. All you gain from switching is an extra dependency. It may look simpler, but that's merely because you push the complexity from your project up to this extra external dependency. Then you have to depend on Meson maintainers to fix any problems you may have and to conform to whatever they decide is the "right" way to build a project.

As for the specific reasons mentioned:

> Portable
> The current build system is not very portable.

I use this build system for all my projects. Over the last fifteen years it has built them on Linux, FreeBSD, SunOS, and Solaris. I don't know what else you want to be portable to.

> It seems to depend on GNU extensions to utilities like sed.

I'm not aware of it using any nonportable extensions. It does, after all, run on all the existing unix-like OSes. Feel free to mention what they are, on which platforms they break the build, and I'll see if their use can be removed.

> Integrated tests

Speaking only for myself, I've never been able to figure out the value of integrated tests for UI. It seems to require you to implement a whole windowing system and user emulation for every test.

> Currently, testing berry would mean writing an original test framework
> to go with the original build system.

Most people use an existing test framework, and it usually is not integrated with the build system. Locking yourself into a framework that is, sounds like a bad idea to me. What will you do when the next fad comes along?

> Simple

Only because you would be relying on Meson to handle all the complex stuff. This is not always a bad thing, but in this case adding it as an extra dependency creates more complexity than it removes.

While the current Makefile/configure are indeed larger, most of the code is not specific to berry. I didn't write it from scratch here, I merely copied it from one of my other projects and adapted a few lines, which took less time that it would take you to write the Meson files from scratch. And my way uses only what is already installed on everyone's system. Hardly anyone uses Meson, and I have seen no reason so far to encourage them to.

> Automatic compile_commands.json

Who cares?

> Implicit behavior
> With Meson, a "make dist" target is not needed.

Another way to put it is that "make dist" is built into Meson. This means that its behavior is "my way or the highway". That may be fine if you like it, but intolerable if you don't. I don't `make dist`, for example, I `make install`, and that is what you'd have to do to get packaged for any distribution.

> Meson build definitions include enough information that Meson knows what to include in a distfile.

In other words, it guesses. What do I do if it guesses wrong? Read Meson source code to figure out how to work around it? If anything goes wrong with the Makefile or the configure script, then it's right there in front of you to do with as you please. And, of course, anybody can hack a simple shell script. Do you have the social skills to argue with Meson developers to let you do what you want to do?

> It's also not necessary to create "release" and "debug" targets

We make the debug build for testing during development. Nobody else needs to.

> Meson automatically makes use of ccache if it's installed.

For 3 .c files?

> Meson automatically figures out the directory to install to

So does configure.

**Comment by abenson** [comment]
I would like it to properly support `DESTDIR`.
